File: CompilerProject_CPP/src/IntermediateCode/IntermediateCodeGenerator.cpp
#include "../../include/IntermediateCode/IntermediateCodeGenerator.h"

namespace CompilerCPP {

    IntermediateCodeGenerator::IntermediateCodeGenerator()
        : _tempCounter(0), _labelCounter(0) {}

    std::string IntermediateCodeGenerator::NewTemp() {
        return "t" + std::to_string(++_tempCounter);
    }

    std::string IntermediateCodeGenerator::NewLabel() {
        return "L" + std::to_string(++_labelCounter);
    }

    std::vector<std::string> IntermediateCodeGenerator::Generate(const std::shared_ptr<Node>& root) {
        _tac.clear();
        _tempCounter = 0;
        _labelCounter = 0;

        if (!root) return _tac;

        _tac.push_back("// ===================================================");
        _tac.push_back("// بداية الكود الوسيط (TAC) للبرنامج: " + root->Name);
        _tac.push_back("// ===================================================");

        if (!root->Children.empty()) {
            auto block = root->Children[0];
            if (!block->Children.empty()) {
                GenerateDeclarations(block->Children[0]);
                if (block->Children.size() > 1) {
                    auto stmtList = block->Children[1];
                    for (const auto& stmt : stmtList->Children) {
                        GenerateStatement(stmt);
                    }
                }
            }
        }

        _tac.push_back("// نهاية البرنامج");
        return _tac;
    }

    void IntermediateCodeGenerator::GenerateDeclarations(const std::shared_ptr<Node>& node) {
        if (!node) return;

        for (const auto& sec : node->Children) {
            if (sec->Value == "ConstDeclarations") {
                for (const auto& c : sec->Children) {
                    _tac.push_back(c->Name + " = " + c->Val);
                }
            }
        }
    }

    void IntermediateCodeGenerator::GenerateStatement(const std::shared_ptr<Node>& node) {
        if (!node) return;

        if (node->Value == "Assign") {
            std::string exprRes = GenerateExpression(node->Children[1]);
            std::string target = node->Name;
            if (node->Children[0]->Value == "FieldAccess") {
                target = node->Children[0]->Children[0]->Name + "." + node->Children[0]->Name;
            } else if (node->Children[0]->Value == "IndexedAccess") {
                std::string idx = GenerateExpression(node->Children[0]->Children[1]);
                target = node->Children[0]->Children[0]->Name + "[" + idx + "]";
            }
            _tac.push_back(target + " = " + exprRes);
        } else if (node->Value == "PrintStatement") {
            for (const auto& arg : node->Children) {
                if (arg->Value == "StringLiteral") {
                    _tac.push_back("print \"" + arg->Val + "\"");
                } else {
                    std::string res = GenerateExpression(arg);
                    _tac.push_back("print " + res);
                }
            }
        } else if (node->Value == "ReadStatement") {
            _tac.push_back("read " + node->Name);
        } else if (node->Value == "IfStatement") {
            std::string cond = GenerateExpression(node->Children[0]);
            std::string trueLabel = NewLabel();
            std::string falseLabel = NewLabel();
            std::string endLabel = NewLabel();

            _tac.push_back("if " + cond + " goto " + trueLabel);
            _tac.push_back("goto " + falseLabel);
            _tac.push_back(trueLabel + ":");

            auto trueBlock = node->Children[1];
            if (trueBlock->Value == "StatementList") {
                for (const auto& s : trueBlock->Children) GenerateStatement(s);
            } else {
                GenerateStatement(trueBlock);
            }

            if (node->Children.size() > 2) {
                _tac.push_back("goto " + endLabel);
                _tac.push_back(falseLabel + ":");

                auto elseBlock = node->Children[2]->Children[0];
                if (elseBlock->Value == "StatementList") {
                    for (const auto& s : elseBlock->Children) GenerateStatement(s);
                } else {
                    GenerateStatement(elseBlock);
                }
                _tac.push_back(endLabel + ":");
            } else {
                _tac.push_back(falseLabel + ":");
            }
        } else if (node->Value == "WhileStatement") {
            std::string startLabel = NewLabel();
            std::string bodyLabel = NewLabel();
            std::string endLabel = NewLabel();

            _tac.push_back(startLabel + ":");
            std::string cond = GenerateExpression(node->Children[0]);
            _tac.push_back("if " + cond + " goto " + bodyLabel);
            _tac.push_back("goto " + endLabel);
            _tac.push_back(bodyLabel + ":");

            auto body = node->Children[1];
            if (body->Value == "StatementList") {
                for (const auto& s : body->Children) GenerateStatement(s);
            } else {
                GenerateStatement(body);
            }

            _tac.push_back("goto " + startLabel);
            _tac.push_back(endLabel + ":");
        } else if (node->Value == "RepeatUntilStatement") {
            std::string startLabel = NewLabel();
            std::string exitLabel = NewLabel();
            _tac.push_back(startLabel + ":");

            auto body = node->Children[0];
            if (body->Value == "StatementList") {
                for (const auto& s : body->Children) GenerateStatement(s);
            } else {
                GenerateStatement(body);
            }

            std::string cond = GenerateExpression(node->Children[1]);
            _tac.push_back("if " + cond + " goto " + exitLabel);
            _tac.push_back("goto " + startLabel);
            _tac.push_back(exitLabel + ":");
        } else if (node->Value == "ForStatement") {
            std::string loopVar = node->Name;
            std::string startVal = GenerateExpression(node->Children[0]);
            std::string endVal = GenerateExpression(node->Children[1]);
            std::string stepVal = (node->Children.size() > 3) ? GenerateExpression(node->Children[2]) : "1";

            _tac.push_back(loopVar + " = " + startVal);
            std::string startLabel = NewLabel();
            std::string bodyLabel = NewLabel();
            std::string endLabel = NewLabel();

            _tac.push_back(startLabel + ":");
            std::string condTemp = NewTemp();
            _tac.push_back(condTemp + " = " + loopVar + " <= " + endVal);
            _tac.push_back("if " + condTemp + " goto " + bodyLabel);
            _tac.push_back("goto " + endLabel);
            _tac.push_back(bodyLabel + ":");

            auto body = node->Children.back();
            if (body->Value == "StatementList") {
                for (const auto& s : body->Children) GenerateStatement(s);
            } else {
                GenerateStatement(body);
            }

            std::string incTemp = NewTemp();
            _tac.push_back(incTemp + " = " + loopVar + " + " + stepVal);
            _tac.push_back(loopVar + " = " + incTemp);
            _tac.push_back("goto " + startLabel);
            _tac.push_back(endLabel + ":");
        }
    }

    std::string IntermediateCodeGenerator::GenerateExpression(const std::shared_ptr<Node>& node) {
        if (!node) return "";

        if (node->Value == "Number" || node->Value == "StringLiteral" || node->Value == "BooleanLiteral" || node->Value == "CharLiteral") {
            return node->Val;
        }

        if (node->Value == "Variable") {
            return node->Name;
        }

        if (node->Value == "FieldAccess") {
            return node->Children[0]->Name + "." + node->Name;
        }

        if (node->Value == "IndexedAccess") {
            std::string idx = GenerateExpression(node->Children[1]);
            return node->Children[0]->Name + "[" + idx + "]";
        }

        if (node->Value == "UnaryExpr") {
            std::string sub = GenerateExpression(node->Children[0]);
            std::string temp = NewTemp();
            _tac.push_back(temp + " = " + node->Name + " " + sub);
            return temp;
        }

        if (node->Value == "BinaryExpr") {
            std::string left = GenerateExpression(node->Children[0]);
            std::string right = GenerateExpression(node->Children[1]);
            std::string temp = NewTemp();
            _tac.push_back(temp + " = " + left + " " + node->Name + " " + right);
            return temp;
        }

        return "";
    }

} // namespace CompilerCPP

File: CompilerProject/IntermediateCode/IntermediateCodeGenerator.cs
using System;
using System.Collections.Generic;
using System.Text;
using CompilerProject.Models;

namespace CompilerProject.IntermediateCode
{
    /// <summary>
    /// مولد الكود الوسيط (Intermediate Code Generator)
    /// يولد كود ثلاثي العناوين (Three-Address Code - TAC) لجميع تراكيب اللغة العربية
    /// Reference: قواعد لغة البرمجة العربية - جامعة إب (ص 3 - 5)
    /// </summary>
    public class IntermediateCodeGenerator
    {
        private int _tempCounter = 1;
        private int _labelCounter = 1;
        private readonly List<string> _instructions = new List<string>();

        public IReadOnlyList<string> Instructions => _instructions;

        private string NewTemp() => $"t{_tempCounter++}";
        private string NewLabel() => $"L{_labelCounter++}";

        /// <summary>
        /// توليد الكود الوسيط من شجرة الإعراب المجردة (AST)
        /// </summary>
        public List<string> Generate(Node rootNode)
        {
            _instructions.Clear();
            _tempCounter = 1;
            _labelCounter = 1;

            if (rootNode == null) return _instructions;

            _instructions.Add("// ===================================================");
            _instructions.Add($"// بداية الكود الوسيط (TAC) للبرنامج: {rootNode.Name}");
            _instructions.Add("// ===================================================");

            GenerateNode(rootNode);

            _instructions.Add("// نهاية البرنامج");
            return _instructions;
        }

        private void GenerateNode(Node node)
        {
            if (node == null) return;

            switch (node.Value)
            {
                case "ProgramRoot":
                case "Block":
                case "Declarations":
                case "ConstDeclarations":
                case "VarDeclarations":
                case "StatementList":
                    foreach (var child in node.Children)
                    {
                        GenerateNode(child);
                    }
                    break;

                case "ConstDecl":
                    _instructions.Add($"{node.Name} = {node.Val}");
                    break;

                case "Assign":
                    GenerateAssign(node);
                    break;

                case "ReadStatement":
                    if (node.Children.Count > 0)
                    {
                        string varName = node.Children[0].Name;
                        _instructions.Add($"read {varName}");
                    }
                    break;

                case "PrintStatement":
                    GeneratePrint(node);
                    break;

                case "IfStatement":
                    GenerateIf(node);
                    break;

                case "WhileStatement":
                    GenerateWhile(node);
                    break;

                case "RepeatUntilStatement":
                    GenerateRepeatUntil(node);
                    break;

                case "ForStatement":
                    GenerateFor(node);
                    break;
            }
        }

        private void GenerateAssign(Node assignNode)
        {
            string varName = assignNode.Name;
            if (assignNode.Children.Count > 1)
            {
                string exprResult = GenerateExpression(assignNode.Children[1]);
                _instructions.Add($"{varName} = {exprResult}");
            }
        }

        private void GeneratePrint(Node printNode)
        {
            foreach (var child in printNode.Children)
            {
                if (child.Value == "StringLiteral")
                {
                    _instructions.Add($"print \"{child.Val}\"");
                }
                else
                {
                    string res = GenerateExpression(child);
                    _instructions.Add($"print {res}");
                }
            }
        }

        private void GenerateIf(Node ifNode)
        {
            if (ifNode.Children.Count < 2) return;

            string trueLabel = NewLabel();
            string falseLabel = NewLabel();
            string endLabel = NewLabel();

            var condNode = ifNode.Children[0];
            var thenNode = ifNode.Children[1];
            Node? elseNode = ifNode.Children.Count > 2 ? ifNode.Children[2] : null;

            // توليد الشرط
            GenerateCondition(condNode, trueLabel, falseLabel);

            // كتلة Then
            _instructions.Add($"{trueLabel}:");
            GenerateNode(thenNode);

            if (elseNode != null)
            {
                _instructions.Add($"goto {endLabel}");
                _instructions.Add($"{falseLabel}:");
                GenerateNode(elseNode);
                _instructions.Add($"{endLabel}:");
            }
            else
            {
                _instructions.Add($"{falseLabel}:");
            }
        }

        private void GenerateWhile(Node whileNode)
        {
            if (whileNode.Children.Count < 2) return;

            string startLabel = NewLabel();
            string bodyLabel = NewLabel();
            string endLabel = NewLabel();

            _instructions.Add($"{startLabel}:");
            GenerateCondition(whileNode.Children[0], bodyLabel, endLabel);

            _instructions.Add($"{bodyLabel}:");
            GenerateNode(whileNode.Children[1]);
            _instructions.Add($"goto {startLabel}");

            _instructions.Add($"{endLabel}:");
        }

        private void GenerateRepeatUntil(Node repeatNode)
        {
            if (repeatNode.Children.Count < 2) return;

            string loopLabel = NewLabel();
            string endLabel = NewLabel();

            _instructions.Add($"{loopLabel}:");
            GenerateNode(repeatNode.Children[0]); // الجسم أولاً

            // الفحص: إذا تحقق الشرط يخرج، وإلا يكرر
            GenerateCondition(repeatNode.Children[1], endLabel, loopLabel);
            _instructions.Add($"{endLabel}:");
        }

        private void GenerateFor(Node forNode)
        {
            string loopVar = forNode.Name;
            string startVal = GenerateExpression(forNode.Children[0]);
            string endVal = GenerateExpression(forNode.Children[1]);
            string stepVal = forNode.Children.Count > 3 ? GenerateExpression(forNode.Children[2]) : "1";
            var bodyNode = forNode.Children[forNode.Children.Count - 1];

            string startLabel = NewLabel();
            string bodyLabel = NewLabel();
            string endLabel = NewLabel();

            // التهيئة الأولية
            _instructions.Add($"{loopVar} = {startVal}");

            _instructions.Add($"{startLabel}:");
            _instructions.Add($"if {loopVar} <= {endVal} goto {bodyLabel}");
            _instructions.Add($"goto {endLabel}");

            _instructions.Add($"{bodyLabel}:");
            GenerateNode(bodyNode);
            _instructions.Add($"{loopVar} = {loopVar} + {stepVal}");
            _instructions.Add($"goto {startLabel}");

            _instructions.Add($"{endLabel}:");
        }

        private void GenerateCondition(Node condNode, string trueLabel, string falseLabel)
        {
            if (condNode.Value == "BinaryExpr")
            {
                if (IsRelationalOp(condNode.Val))
                {
                    string left = GenerateExpression(condNode.Children[0]);
                    string right = GenerateExpression(condNode.Children[1]);
                    _instructions.Add($"if {left} {condNode.Val} {right} goto {trueLabel}");
                    _instructions.Add($"goto {falseLabel}");
                    return;
                }

                if (condNode.Val == "&&")
                {
                    string nextLabel = NewLabel();
                    GenerateCondition(condNode.Children[0], nextLabel, falseLabel);
                    _instructions.Add($"{nextLabel}:");
                    GenerateCondition(condNode.Children[1], trueLabel, falseLabel);
                    return;
                }

                if (condNode.Val == "||")
                {
                    string nextLabel = NewLabel();
                    GenerateCondition(condNode.Children[0], trueLabel, nextLabel);
                    _instructions.Add($"{nextLabel}:");
                    GenerateCondition(condNode.Children[1], trueLabel, falseLabel);
                    return;
                }
            }

            string res = GenerateExpression(condNode);
            _instructions.Add($"if {res} != 0 goto {trueLabel}");
            _instructions.Add($"goto {falseLabel}");
        }

        private string GenerateExpression(Node exprNode)
        {
            if (exprNode == null) return "";

            if (exprNode.Value == "Number") return exprNode.Val;
            if (exprNode.Value == "Variable") return exprNode.Name;
            if (exprNode.Value == "BooleanLiteral") return exprNode.Val == "صح" ? "1" : "0";
            if (exprNode.Value == "StringLiteral") return $"\"{exprNode.Val}\"";
            if (exprNode.Value == "CharLiteral") return $"'{exprNode.Val}'";

            if (exprNode.Value == "FieldAccess")
            {
                string target = exprNode.Children.Count > 0 ? GenerateExpression(exprNode.Children[0]) : "obj";
                return $"{target}.{exprNode.Name}";
            }

            if (exprNode.Value == "IndexedAccess")
            {
                string target = exprNode.Children.Count > 0 ? GenerateExpression(exprNode.Children[0]) : "arr";
                string idx = exprNode.Children.Count > 1 ? GenerateExpression(exprNode.Children[1]) : "0";
                return $"{target}[{idx}]";
            }

            if (exprNode.Value == "UnaryExpr")
            {
                string operand = GenerateExpression(exprNode.Children[0]);
                string t = NewTemp();
                _instructions.Add($"{t} = {exprNode.Val} {operand}");
                return t;
            }

            if (exprNode.Value == "BinaryExpr")
            {
                string left = GenerateExpression(exprNode.Children[0]);
                string right = GenerateExpression(exprNode.Children[1]);
                string t = NewTemp();
                _instructions.Add($"{t} = {left} {exprNode.Val} {right}");
                return t;
            }

            return exprNode.Name;
        }

        private static bool IsRelationalOp(string op)
        {
            return op is "==" or "!=" or "<" or ">" or "<=" or ">=";
        }
    }
}
